error-prone-contrib/src/main/java/tech/picnic/errorprone/bugpatterns/MonoZipOfMonoVoidUsage.java

```diff
@@ -0,0 +1,161 @@
+import static com.google.errorprone.matchers.Matchers.toType;
+import static tech.picnic.errorprone.bugpatterns.util.Documentation.BUG_PATTERNS_BASE_URL;
+import static tech.picnic.errorprone.bugpatterns.util.MoreTypes.generic;
+import static tech.picnic.errorprone.bugpatterns.util.MoreTypes.type;
+
+import com.google.auto.service.AutoService;
+import com.google.common.collect.Iterables;
+import com.google.errorprone.BugPattern;
+import com.google.errorprone.VisitorState;
+import com.google.errorprone.bugpatterns.BugChecker;
+import com.google.errorprone.bugpatterns.BugChecker.MethodInvocationTreeMatcher;
+import com.google.errorprone.fixes.SuggestedFixes;
+import com.google.errorprone.matchers.Description;
+import com.google.errorprone.matchers.Matcher;
+import com.google.errorprone.suppliers.Supplier;
+import com.google.errorprone.util.ASTHelpers;
+import com.sun.source.tree.ExpressionTree;
+import com.sun.source.tree.MemberSelectTree;
+import com.sun.source.tree.MethodInvocationTree;
+import com.sun.tools.javac.code.Type;
+import java.util.Optional;
+
+/**
+ * A {@link BugChecker} that flags usages of {@link
+ * reactor.core.publisher.Mono#zip(reactor.core.publisher.Mono, reactor.core.publisher.Mono)} and
+ * {@link reactor.core.publisher.Mono#zipWith(reactor.core.publisher.Mono)} with {@link
+ * reactor.core.publisher.Mono#empty()} parameters.
+ *
+ * <p>{@link reactor.core.publisher.Mono#zip(reactor.core.publisher.Mono,
+ * reactor.core.publisher.Mono)} and {@link
+ * reactor.core.publisher.Mono#zipWith(reactor.core.publisher.Mono)} perform incorrectly upon
+ * retrieval of the empty publisher and prematurely terminates the reactive chain from the
+ * execution. In most cases this is not the desired behaviour.
+ *
+ * <p>NB: {@code Mono<?>.zipWith(Mono<Void>)} is allowed by the Reactor API, but it is an incorrect
+ * usage of the API. It will be flagged by ErrorProne but the fix won't be supplied. The problem
+ * with the original code should be revisited and fixed in a structural manner by the developer.
+ */
+@AutoService(BugChecker.class)
+@BugPattern(
+    summary =
+        "`Mono#zip` and `Mono#zipWith` should not be executed against `Mono#empty` or `Mono<Void>` parameter; "
+            + "please revisit the parameters used and make sure to supply correct publishers instead",
+    link = BUG_PATTERNS_BASE_URL + "MonoZipOfMonoVoidUsage",
+    linkType = CUSTOM,
+    severity = ERROR,
+    tags = LIKELY_ERROR)
+public final class MonoZipOfMonoVoidUsage extends BugChecker
+    implements MethodInvocationTreeMatcher {
+  private static final long serialVersionUID = 1L;
+  private static final Supplier<Type> MONO = type("reactor.core.publisher.Mono");
+
+  /**
+   * In fact, we use {@code Mono<Void>} everywhere in codebases instead of {@code Mono<Object>}
+   * (actual return type of {@code Mono.empty()}) to represent empty publisher.
+   */
+  private static final Supplier<Type> MONO_VOID_TYPE =
```

Comment:
WDYT about ?
```suggestion
  private static final Supplier<Type> MONO_VOID_TYPE_SUPPLIER =
```

Reply:
Also fine by me. Will change other one accordingly

Reply:
@mohamedsamehsalah I don't think that suggestion is consistent with the rest of the codebase:
```
$ git grep 'private static final Supplier'
error-prone-contrib/src/main/java/tech/picnic/errorprone/bugpatterns/AssociativeMethodInvocation.java:  private static final Supplier<Type> ITERABLE = Suppliers.typeFromClass(Iterable.class);
error-prone-contrib/src/main/java/tech/picnic/errorprone/bugpatterns/FluxFlatMapUsage.java:  private static final Supplier<Type> FLUX =
error-prone-contrib/src/main/java/tech/picnic/errorprone/bugpatterns/FluxFlatMapUsage.java:  private static final Supplier<Type> FLUX_OF_PUBLISHERS =
error-prone-contrib/src/main/java/tech/picnic/errorprone/bugpatterns/FluxImplicitBlock.java:  private static final Supplier<Type> STREAM =
error-prone-contrib/src/main/java/tech/picnic/errorprone/bugpatterns/NestedOptionals.java:  private static final Supplier<Type> OPTIONAL = Suppliers.typeFromClass(Optional.class);
error-prone-contrib/src/main/java/tech/picnic/errorprone/bugpatterns/NestedPublishers.java:  private static final Supplier<Type> PUBLISHER = type("org.reactivestreams.Publisher");
error-prone-contrib/src/main/java/tech/picnic/errorprone/bugpatterns/StringJoin.java:  private static final Supplier<Type> CHAR_SEQUENCE_TYPE =
error-prone-contrib/src/main/java/tech/picnic/errorprone/bugpatterns/StringJoin.java:  private static final Supplier<Type> FORMATTABLE_TYPE = Suppliers.typeFromClass(Formattable.class);
refaster-runner/src/main/java/tech/picnic/errorprone/refaster/runner/CodeTransformers.java:  private static final Supplier<ImmutableListMultimap<String, CodeTransformer>>
```

(In this context the `Supplier` wrapper is mostly a performance optimization; adding it to the name doesn't really increase readability.)